# Control Statements and Program Development

## Exercises

3.13 (_Perfect Numbers_)

In [1]:
def list_primes(n):
    """
    This function lists all primes least than or equal to a given integer number 'n'.
    """
    if n < 2:
        return []
    primes = list(list(range(2, n+1)))
    limit = 1+ int(n**(0.5))
    for i in range(limit):
        primes = [_ for _ in primes if ((_ % primes[i] != 0) or (_ <= primes[i]))]
    return primes 

In [2]:
import time
for n in range(1000, 10001, 1000):
    start_time = time.perf_counter()
    list_primes(n)
    end_time = time.perf_counter() - start_time
    print(f"{n:>5d} {end_time:>.5f}")

 1000 0.00309
 2000 0.00828
 3000 0.01109
 4000 0.02234
 5000 0.02800
 6000 0.02818
 7000 0.03635
 8000 0.04221
 9000 0.04902
10000 0.06148


In [3]:
def list_divisors(n):
    """
    This function lists the divisors of a given integer 'n'
    """
    limit = 1 + n//2
    divisors = [_ for _ in range(1, limit) if n % _ == 0 ]
    return divisors

In [4]:
def is_perfect(n):
    """
    This function returns 'True' if 'n' is a perfect number, or 'False', otherwise.
    """
    if n == sum(list_divisors(n)):
        return True
    else:
        return False

In [5]:
for n in range(1,10000):
    if is_perfect(n):
        print(f"{n:>4d}")

   6
  28
 496
8128


3.14 (_Challenge: Approximation the Mathematical Constant $\pi$_)

In [6]:
def pi(n = 10):
    pi = 4
    for i in range(1, n+1):
        pi += (-1)**(i) * 4 / (2 * i + 1)
    return pi

In [7]:
print(10 * "=")
print(f"{'n':>3s} {'pi':>6s}")
print(10 * "-")
for i in range(1,11):
    print(f"{i:>3d} {pi(i):>6.4f}")
print(10 * "=")

  n     pi
----------
  1 2.6667
  2 3.4667
  3 2.8952
  4 3.3397
  5 2.9760
  6 3.2837
  7 3.0171
  8 3.2524
  9 3.0418
 10 3.2323


3.17 (_Nested Loops_)

In [15]:
for i in range(1,11):
    for j in range(i-1):
        print('*', end='')
    print('*')

*
**
***
****
*****
******
*******
********
*********
**********


In [19]:
for i in range(10, 0, -1):
    for j in range(i-1):
        print('*', end='')
    print('*')

**********
*********
********
*******
******
*****
****
***
**
*


In [29]:
for i in range(10):
    for j in range(i):
        print(' ', end='')
    for k in range(10, i+1, -1):
        print('*', end='')
    print('*')

**********
 *********
  ********
   *******
    ******
     *****
      ****
       ***
        **
         *


In [30]:
for i in range(10):
    for j in range(10, i+1, -1):
        print(' ', end='')
    for k in range(i):
        print('*', end='')
    print('*')

         *
        **
       ***
      ****
     *****
    ******
   *******
  ********
 *********
**********


3.19 (_Prime Factorization_)

In [36]:
def prime_factorization(n):
    primes = list_primes(n)
    if n in primes:
        return [[n, 1]]
    factors = []
    while n != 1:
        for p in primes:
            t = [p, 0]
            while n % p == 0:
                t[1] += 1
                n /= p
            if t[1] > 0:
                factors.append(t)
    return factors

In [41]:
prime_factorization(1537)

[[29, 1], [53, 1]]

3.28 (_Intro to Data Science: Mean, Median and Mode_)

In [47]:
temperatures = [19.5, 19.5, 21.6, 20.2, 18.6, 17.2, 19.5]

In [49]:
mean = sum(temperatures) / len(temperatures)
mean

19.442857142857147

In [51]:
temperatures.sort()
n = len(temperatures)
if n % 2 == 0:
    median = (temperatures[n//2] + temperatures[n//2 + 1])/2
else:
    median = temperatures[n//2 + 1]
median

19.5

In [52]:
temperatures

[17.2, 18.6, 19.5, 19.5, 19.5, 20.2, 21.6]

In [56]:
temperatures = [1, 1, 1, 19.5, 19.5, 21.6, 21.6, 21.6, 20.2, 18.6, 17.2, 19.5]

In [64]:
unique_temp = []
qnt_temp = []
for temp in temperatures:
    if temp not in unique_temp:
        unique_temp.append(temp)
        qnt_temp.append(1)
    else:
        index = unique_temp.index(temp)
        qnt_temp[index] += 1

mode = []
q_id = qnt_temp.index(max(qnt_temp))
mode = [t for q, t in zip(qnt_temp, unique_temp) if q == q_id]
mode

[]